### PR TITLE
fix: prevent eager-loaded entities with DeleteDateColumn from overwriting manual relations

### PR DESCRIPTION
**Pull request by LeviHeber**
The issue was caused by the way TypeORM processing queries. For each join, it checks if the join for that table exists. However, if the join has a condition, it will always add the join, even if a previous join for the same table already exists.

In the case of eager loading, TypeORM creates its own join, and since it has a DeleteDateColumn, an additional condition is added to the join (checking for a null DeletedDate). This newly added join does not match the manually specified join, and if the manual relation had additional nested fields that were not part of the eager load, those fields would be overwritten.

This fix addresses the issue by merging the manually requested relations with the eager-loaded data, preserving the complete details of the manual relations.

In issue #10426 , the same case was described, where there are 2 joins to the same table,
which causes performance issues.

The current fix does not change the join process and leaves both joins as they were,
but only fixes the consequences that were created as a result of it in the object creation,
because this is a consequence that is relevant both to this case and to other cases.

Unrelated, it should check how to deal with the problem of having 2 joins, for performance reasons.

Closes: #11265
Closes: #11056
Closes: #10812
Closes: #9774
Closes: #9581

### Pull-Request Checklist

- [X] Code is up-to-date with the `master` branch
- [X] `npm run format` to apply prettier formatting
- [X] `npm run test` passes with this change
- [X] This pull request links relevant issues as `Fixes #0000`
- [X] There are new or updated unit tests validating the change
- [ ] N/A - Documentation has been updated to reflect this change
- [X] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Review by gioboa — COMMENTED** [review]
Hi @LeviHeber, thanks for this great PR, can you fix up the sap tests please?

File: test/github-issues/11265/issue-11265.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        return Promise.all(
+            dataSources.map(async (dataSource) => {
+                // Prepare test data
+                const relationNestedRepository =
+                    dataSource.getRepository(RelationNestedEntity)
+                const relationEntityRepository =
+                    dataSource.getRepository(RelationEntity)
+                const parentRepository = dataSource.getRepository(ParentEntity)
+
+                const relationNested = new RelationNestedEntity()
```
Comment (alumni) [inline test/github-issues/11265/issue-11265.ts]: SAP HANA cannot insert a row without values (there must be at least one value that is not auto-generated).

Comment (LeviHeber) [reply test/github-issues/11265/issue-11265.ts]: I added a non auto-generated dummy field.

**Comment by LeviHeber** [comment]
i left the test to be `describe.only`. is it need to be fixed?

**Review by gioboa — COMMENTED** [review]
Yes, remove the .only please

**Comment by LeviHeber** [comment]
I removed the `only` and pushed it again.
It seems that more tests are running now, but for some reason many of the tests encounter the same check (`commands - migration create: should write Javascript empty migration file when option is passed`) and it doesn't seem to be related to the fix I made.
I would appreciate some guidance.

**Comment by sgarner** [comment]
> I removed the `only` and pushed it again. It seems that more tests are running now, but for some reason many of the tests encounter the same check (`commands - migration create: should write Javascript empty migration file when option is passed`) and it doesn't seem to be related to the fix I made. I would appreciate some guidance.

That sounds possibly related to #11653 which has been fixed now. Tests are passing in this PR.

File: test/github-issues/11265/issue-11265.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    closeTestingConnections,
+    reloadTestingDatabases,
+} from "../../utils/test-utils"
+import { DataSource } from "../../../src/data-source/DataSource"
+import { expect } from "chai"
+import { RelationNestedEntity } from "./entity/relationNestedEntity"
+import { RelationEntity } from "./entity/relationEntity"
+import { ParentEntity } from "./entity/parent"
+
+describe("github issues > #11265 Eager Relations with DeleteDateColumn Overwrite Manual Relations During Query Loading", () => {
```
Comment (sgarner) [inline test/github-issues/11265/issue-11265.ts]: We aren't creating tests for specific GitHub issues any more. Instead, tests should be added to the test suite(s) for the related features.

I suggest moving this test to `test/functional/relations/eager-relations/basic-eager-relations.ts`

File: src/metadata/RelationMetadata.ts [hunk 3]
```diff
@@ -520,6 +520,8 @@ export class RelationMetadata {
                 entity,
             )
         } else {
+            if (ObjectUtils.isObject(entity[propertyName]))
+                ObjectUtils.assign(value, entity[propertyName])
             entity[propertyName] = value
```
Comment (sgarner) [inline src/metadata/RelationMetadata.ts]: If I'm reading it right, the proposed change will only augment `entity[propertyName]` with _additional_ fields from `value`. Any fields already existing in `entity[propertyName]` won't be modified.

Also note that `ObjectUtils.assign()` mutates the target object. Depending on the `value` being passed, this might have unintended consequences.

Should we be merging objects deeply here? Is it possible for there to be two joins with overlapping _nested_ relations? 

```suggestion
            if (ObjectUtils.isObject(entity[propertyName])) {
                entity[propertyName] = OrmUtils.mergeDeep(entity[propertyName], value)
            } else {
                entity[propertyName] = value
            }
```

**Comment by LeviHeber** [comment]
I made the following fixes:
1. Moved the test to its proper place in the eager-relations feature tests.
2. Updated the fix to work with `OrmUtils.mergeDeep` instead of `ObjectUtils.assign`

**Comment by LeviHeber** [comment]
Is there anything else that needs to be done to move forward?

**Comment by naorpeled** [comment]
@coderabbitai review

**Review by gioboa — APPROVED** [review]

**Comment by alumni** [comment]
I think the issue that we have could possibly be a bit earlier, when we build the query. It seems we select the user twice - maybe once because it's an eagerly loaded relation and another time because it was requested by the user.

In the test we have:

```ts
await connection.manager.findOne(
    Editor,
    {
        where: { userId: 1 },
        relations: {
            user: {
                nestedProfile: true,
            },
        },
    },
)
```

which generates the query:

```sql
SELECT
    "Editor"."userId" AS "Editor_userId",
    "Editor"."postId" AS "Editor_postId",
    "Editor__Editor_user"."id" AS "Editor__Editor_user_id",
    "Editor__Editor_user"."firstName" AS "Editor__Editor_user_firstName",
    "Editor__Editor_user"."lastName" AS "Editor__Editor_user_lastName",
    "Editor__Editor_user"."deletedAt" AS "Editor__Editor_user_deletedAt",
    "Editor__Editor_user"."profileId" AS "Editor__Editor_user_profileId",
    "Editor__Editor_user"."nestedProfileId" AS "Editor__Editor_user_nestedProfileId",
    "Editor__Editor_user__Editor__Editor_user_nestedProfile"."id" AS "Editor__Editor_user__Editor__Editor_user_nestedProfile_id",
    "Editor__Editor_user__Editor__Editor_user_nestedProfile"."about" AS "Editor__Editor_user__Editor__Editor_user_nestedProfile_about",
    "Editor__Editor_user__Editor__Editor_user_profile"."id" AS "Editor__Editor_user__Editor__Editor_user_profile_id",
    "Editor__Editor_user__Editor__Editor_user_profile"."about" AS "Editor__Editor_user__Editor__Editor_user_profile_about",
    "Editor__user"."id" AS "Editor__user_id",
    "Editor__user"."firstName" AS "Editor__user_firstName",
    "Editor__user"."lastName" AS "Editor__user_lastName",
    "Editor__user"."deletedAt" AS "Editor__user_deletedAt",
    "Editor__user"."profileId" AS "Editor__user_profileId",
    "Editor__user"."nestedProfileId" AS "Editor__user_nestedProfileId",
    "Editor__user__profile"."id" AS "Editor__user__profile_id",
    "Editor__user__profile"."about" AS "Editor__user__profile_about"
FROM
    "editor" "Editor"
    LEFT JOIN "user" "Editor__Editor_user" ON "Editor__Editor_user"."id" = "Editor"."userId"
    AND ("Editor__Editor_user"."deletedAt" IS NULL)
    LEFT JOIN "profile" "Editor__Editor_user__Editor__Editor_user_nestedProfile" ON "Editor__Editor_user__Editor__Editor_user_nestedProfile"."id" = "Editor__Editor_user"."nestedProfileId"
    LEFT JOIN "profile" "Editor__Editor_user__Editor__Editor_user_profile" ON "Editor__Editor_user__Editor__Editor_user_profile"."id" = "Editor__Editor_user"."profileId"
    LEFT JOIN "user" "Editor__user" ON "Editor__user"."id" = "Editor"."userId"
    AND ("Editor__user"."deletedAt" IS NULL)
    LEFT JOIN "profile" "Editor__user__profile" ON "Editor__user__profile"."id" = "Editor__user"."profileId"
WHERE
    ((("Editor"."userId" = 1)))
    AND (
        "Editor"."userId" = 1
        AND "Editor"."postId" = 1
    )
```

We have both `Editor__Editor_user`+`Editor__Editor_user__Editor__Editor_user_profile` and `Editor__user`+`Editor__user__profile`.

I'll check tomorrow to see if this is intended.

**Comment by alumni** [comment]
We'll create a separate issue for the extra things in the join so we don't block this.

Later edit: created #11823.
